Visual line mode

svetelna opened the pull request:
This PR introduces visual mode line for both webengine and webkit backends. It follows the behaviour of visual line mode in Vim, except for commands move-to-next-word, move-to-prev-word and move-to-end-of-word. It is because in browser is not possible to have cursor even under selection, thus the functionality then will be very similar to move-to-next-line and move-to-prev-line.
Please correct me if otherwise.
In this PR also unit tests are implemented, please let me know me, if more tests are needed.

The-Compiler commented on tests/unit/browser/test_caret.py:
```diff
@@ -61,16 +63,24 @@ def check(self, expected, *, strip=False):
+    def toggle_line(self):
+        with self._qtbot.wait_signal(self._caret.line_selection_toggled):
+            self._caret.toggle_line_selection()
+
```
Similar to above, this could be `def toggle(self, line=False):`

The-Compiler commented on tests/unit/browser/test_caret.py:
```diff
@@ -405,3 +395,82 @@ def test_reverse(self, caret, selection):
+
+class TestLineSelection:
+
+    def test_line_selection_toggle(self, caret, selection):
```
nitpick: No need to repeat the `line_selection` part in every test name since they are already in a `TestLineSelection` class.

svetelna replied:
good point :)

The-Compiler commented on tests/unit/browser/test_caret.py:
```diff
@@ -405,3 +395,82 @@ def test_reverse(self, caret, selection):
+
+                four five six
+                vier fünf sechs
+                """, strip=True)
```
nitpick: Can you align this (and the others using `strip=True`) like the one on lines 460-463, so that the text is indented by one level, and the closing `"""` is on the same level as the function call?

svetelna commented on doc/help/commands.asciidoc:
```diff
@@ -285,8 +285,7 @@ Set all settings back to their default.
 === config-cycle
-Syntax: +:config-cycle [*--pattern* 'pattern'] [*--temp*] [*--print*]
-             'option' ['values' ['values' ...]]+
+Syntax: +:config-cycle [*--pattern* 'pattern'] [*--temp*] [*--print*] 'option' ['values' ['values' ...]]+
```
I only have Python 3.8 installed, but I can remove the docs-update commit, if that would be better.

The-Compiler replied:
Don't worry about it then - I'll regenerate the docs with Python 3.7 (and hopefully find out what is causing this with 3.8) when merging this.

svetelna commented on tests/unit/browser/test_caret.py:
```diff
@@ -61,16 +63,24 @@ def check(self, expected, *, strip=False):
             self._qtbot.wait(50)
 
+        assert False, 'Failed to get selection!'
+
```
I wanted to ask about that, I also noticed failing tests on Travis. But unfortunatelly I do not have access to any Mac computer to investigate it. One test on MacOS is also failing because of line-selection (bug?). I will try to find out why, but as I already mentioned, can not guarantee to find a solution.

svetelna commented:
@The-Compiler Thank you for all the comments, I will try to fix the findings.

svetelna commented:
I have processed the change requests and updated the code. I removed the documentation commit, because of that incorrect docs generation with python 3.8. Unfortunately I did not manage to find out why the tests on Mac platform are not passing (as I can not reproduce it on my computer).
Pylint now shows one warning:
```
qutebrowser/browser/webkit/webkittab.py:767:4: C0116: Missing function or method docstring (missing-function-docstring)
```
But I dont see why specifically this method should have the docstring. Should it?

The-Compiler commented on qutebrowser/browser/browsertab.py:
```diff
@@ -427,13 +427,20 @@ def reapply(self) -> None:
+    'NONE',
+    'NORMAL',
+    'LINE',
+])
```
We use lower-case enum member names usually (see e.g. `qutebrowser/utils/usertypes.py` for some examples).

Also, can you please use the class-based API to create enums instead? See #5336.

The-Compiler commented on qutebrowser/browser/browsertab.py:
```diff
@@ -427,13 +427,20 @@ def reapply(self) -> None:
     """Attribute ``caret`` of AbstractTab for caret browsing."""
 
     #: Signal emitted when the selection was toggled.
     #: (argument - whether the selection is now active)
```
This is outdated now (probably fine to just remove this line, as the type already makes it clear what it is)

The-Compiler commented on qutebrowser/browser/browsertab.py:
```diff
@@ -826,6 +833,19 @@ def networkaccessmanager(self) -> typing.Optional[QNetworkAccessManager]:
+    def run_js_sync(
+            self,
+            code: str, *,
+            world: typing.Union[usertypes.JsWorld, int] = None
```
Since it's only implemented for QtWebKit which can't handle the `world` argument, no need to have it here.

The-Compiler commented on qutebrowser/browser/webengine/webenginetab.py:
```diff
@@ -1225,6 +1233,10 @@ def shutdown(self):
         self._widget.shutdown()
 
+    def run_js_sync(self, code, *, world=None):
+        # pylint: disable=unused-argument
```
You can use `utils.unused(code)` instead.

The-Compiler commented on qutebrowser/browser/webkit/webkittab.py:
```diff
@@ -193,159 +196,198 @@ def _on_mode_entered(self, mode):
-            if self.selection_enabled and self._widget.hasSelection():
+            if self.selection_state is not \
+                    browsertab.SelectionState.NONE and \
+                    self._widget.hasSelection():
```
Please use parentheses around the condition instead, then you won't need to use backslashes for line continuation.

The-Compiler commented on qutebrowser/browser/webkit/webkittab.py:
```diff
@@ -693,6 +764,13 @@ def event_target(self):
 
+    def run_js_sync(self, code, *, world=None):
+        if world is not None and world != usertypes.JsWorld.jseval:
+            log.webview.warning("Ignoring world ID {}".format(world))
```
The `world` argument only exists so that `run_js_async` has the same API with QtWebKit and QtWebEngine - so you can move this check there and remove the `world` argument (as per the comment in `browsertab.py` above)

The-Compiler commented on qutebrowser/components/caretcommands.py:
```diff
@@ -185,9 +185,14 @@ def move_to_end_of_document(tab: apitypes.Tab) -> None:
-def toggle_selection(tab: apitypes.Tab) -> None:
-    """Toggle caret selection mode."""
-    tab.caret.toggle_selection()
+@cmdutils.argument('line')
```
No need for a `@cmdutils.argument` decorator if you don't need to specify any additional information about the argument.

The-Compiler commented on qutebrowser/components/caretcommands.py:
```diff
@@ -185,9 +185,14 @@ def move_to_end_of_document(tab: apitypes.Tab) -> None:
+    """Toggle caret selection mode.
+
+    Args:
+        line: enables line-selection
```
nitpick:

```suggestion
        line: Enables line-selection.
```

The-Compiler commented on qutebrowser/browser/webkit/webkittab.py:
```diff
@@ -693,6 +764,13 @@ def event_target(self):
     def shutdown(self):
         self._widget.shutdown()
 
+    def run_js_sync(self, code, *, world=None):
```
The reason pylint wants a docstring here is because:

- The method is public
- The method is longer than 3 lines

Might be good opportunity to point out that this is QtWebKit-specific here as well.

The-Compiler commented on qutebrowser/browser/webengine/webenginetab.py:
```diff
@@ -1225,6 +1233,10 @@ def shutdown(self):
 
+    def run_js_sync(self, code, *, world=None):
+        # pylint: disable=unused-argument
+        return None
```
This should `raise browsertab.UnsupportedOperationError` instead, like `clear_ssl_errors` above (and probably the docstring should be adjusted accordingly).

For `networkaccessmanager`, the code using it actually needs `None` as a value to say "there is no NAM available", but for `run_js_sync`, `None` could be a valid return value (and not all callers check the return value), so that could lead to subtle bugs when someone isn't aware that it's QtWebKit-specific.

The-Compiler commented:
Cancelling the build as I messed up some things in `master` (sorry!) and I'm waiting for the CI for #5403 - I'll let you know once that's merged, please rebase on master (or merge `master`) again at that point.

The-Compiler commented:
Should be fixed now - can you please update from `master` again - sorry for the trouble!

svetelna commented:
Sure, it's not a problem.

svetelna commented:
Hello, is there something else I need to fix or change? I have already rebased when I was asked to and previous change requests are also processed.

The-Compiler commented:
Ah, I didn't realize your force-push wasn't just a rebase but also added new commits. Personally I prefer merging from master for PRs which already have been reviewed, as that makes it easier for everyone to see what's been going on, compared to force pushes. No biggie though!

There's still one unadressed comment, but it's not a big issue - I can fix that one up while merging if you prefer. I'll still need to take a look at the failing macOS tests though, hopefully will get a chance to do so later this week or early next week.

The-Compiler commented on qutebrowser/javascript/caret.js:
```diff
@@ -1145,16 +1159,52 @@ window._qutebrowser.caret = (function() {
+        if (CaretBrowsing.selectionState === CaretBrowsing.SelectionState.NONE) {
+            return false;
+        }
+        return true;
```
This function could probably just be `return CaretBrowsing.selectionState !== CaretBrowsing.SelectionState.NONE;

The-Compiler replied:
Or alternatively, get rid of the function and use `selectionState !== CaretBrowsing.SelectionState.NONE` directly, like you've done in some (but not all) places below.

svetelna replied:
Sorry I probably missed it somehow, I will fix it.

The-Compiler replied:
No problem!

The-Compiler commented on qutebrowser/javascript/caret.js:
```diff
@@ -1145,16 +1159,52 @@ window._qutebrowser.caret = (function() {
     CaretBrowsing.move = function(direction, granularity, count = 1) {
         let action = "move";
-        if (CaretBrowsing.selectionEnabled) {
+        if (CaretBrowsing.selectionState !== CaretBrowsing.SelectionState.NONE) {
```
Why not `selectionEnabled()`?

svetelna commented:
I should have left some comment with force push then. Nevertheless I have corrected last error and avoided the forcepush, hope its better now.
